docs/architecture/adr-038-state-listening.md

```diff
@@ -204,60 +204,158 @@ func (rs *Store) CacheMultiStore() types.CacheMultiStore {
-    // Listeners returns the streaming service's listeners for the BaseApp to register
-    Listeners() map[types.StoreKey][]store.WriteListener
-    // ABCIListener interface for hooking into the ABCI messages from inside the BaseApp
-    ABCIListener
-    // Closer interface
-    io.Closer
+	// ListenBeginBlock updates the streaming service with the latest BeginBlock messages
+	ListenBeginBlock(blockHeight int64, req []byte, res []byte) error
+	// ListenEndBlock updates the steaming service with the latest EndBlock messages
+	ListenEndBlock(blockHeight int64, req []byte, res []byte) error
+	// ListenDeliverTx updates the steaming service with the latest DeliverTx messages
+	ListenDeliverTx(blockHeight int64, req []byte, res []byte) error
+	// ListenStoreKVPair updates the steaming service with the latest StoreKVPair messages
+	ListenStoreKVPair(blockHeight int64, data []byte) error
+}
+
+// StreamingService struct for registering WriteListeners with the BaseApp and updating the service with the ABCI messages using the hooks
+type StreamingService struct {
+	// Listeners returns the streaming service's listeners for the BaseApp to register
+	Listeners map[types.StoreKey][]store.WriteListener
+	// ABCIListener interface for hooking into the ABCI messages from inside the BaseApp
+	ABCIListener ABCIListener
+	// StopNodeOnErr stops the node when true
+	StopNodeOnErr bool
+}
+```
+
+We will also introduce a new `KVStoreWriter` struct as the writer for use in `StoreKVPairWriteListener`. `KVStoreWriter` will expose `ABCIListener` for streaming state change events via `ABCIListener.ListenStoreKVPair`, a `BlockHeight` func for determining the block height for which the state change events belong to and a boolean flag for determine whether the node needs to be stopped or not.
+
+```go
+// KVStoreWriter is used so that we do not need to update the underlying
+// io.Writer inside the StoreKVPairWriteListener everytime we begin writing
+type KVStoreWriter struct {
+	blockHeight   func() int64
+	listener      ABCIListener
+	stopNodeOnErr bool
+}
+
+// NewKVStoreWriter create an instance of a KVStoreWriter that sends StoreKVPair data to listening service
+func NewKVStoreWriter(
+	listener ABCIListener,
+	stopNodeOnErr bool,
+	blockHeight func() int64,
+) *KVStoreWriter {
+	return &KVStoreWriter{
+		listener:      listener,
+		stopNodeOnErr: stopNodeOnErr,
+		blockHeight:   blockHeight,
+	}
+}
+
+// Write satisfies io.Writer
+func (iw *KVStoreWriter) Write(b []byte) (int, error) {
+	blockHeight := iw.blockHeight()
+	if err := iw.listener.ListenStoreKVPair(blockHeight, b); err != nil {
+		if iw.stopNodeOnErr {
+			panic(err)
+		}
+		return 0, err
+	}
+	return len(b), nil
 }
 ```
 
-#### BaseApp registration
+We will expose a `RegisterStreamingService` function for the App to register streaming services.
+```go
+// RegisterStreamingService registers the ABCI streaming service provided by the streaming plugin.
+func RegisterStreamingService(
```

Comment:
I think all the go-plugins are under same streaming service?
Do we still support multiple streaming services, it seems not the case with the draft implementation.
And if we support multiple streaming services, does each one of them can have different `stop-node-on-err` options?

Reply:
I was looking at this over the weekend and will be my focus today. It does not correctly represent registering multiple streaming services. As it stands this would better be renamed to `RegisterABCIListenerService`. Creating and registering new services should be independent and controlled in app or baseapp. It would better align with the plugin system. So I'm thinking something like:

```
[streaming]

enable = ""

stop-node-on-err = true

# ABCI streaming service
[streaming.abci_service]

keys = ""
plugin = ""

# Some other part of the system
[streaming.other_service]

plugin = ""
```

Reply:
addressed. Check latest updates.